core/templates/expressions/expression-syntax-tree.service.ts

```diff
@@ -23,6 +23,10 @@ import { AppConstants } from 'app.constants';
 import { ExpressionParserService } from
   'expressions/expression-parser.service.ts';
 
+interface Env {
+  [param: string]: Object;
```

Comment:
No, there are `Objects` passed as the `ENV`.

Reply:
@nishantwrp please provide evidence (links to parts of the code).

Reply:
Here 
https://github.com/oppia/oppia/blob/4a0850d03e10d70a7bd8024e874e4a35d3f38811/core/templates/expressions/expression-syntax-tree.service.ts#L84

`this.system` is is treated like a Env. And it is a dict of objects.
https://github.com/oppia/oppia/blob/4a0850d03e10d70a7bd8024e874e4a35d3f38811/core/templates/expressions/expression-syntax-tree.service.ts#L252-L261

Also like outside this file they `strings` are also used like @brianrodri mentioned. So, I think here we could use a generic type `Object`.

Reply:
Sorry @nishantwrp, I'm a bit confused by your reply.

- The code snippets you gave do not actually help. The first one doesn't prove that this.system is a dict of objects (though I can see that it is treated like an Env).
- I don't understand the relevance of the second code example, which has neither reference to Env or this.system.

For your reference, I took a look at the code and I thought that lookupEnvs was a good example of a method that would have answered my question -- notice the `env[name]` construct.

I would like to understand the other part though. My understanding is that env is always a dict, which should be a non-string. So could you list the string cases? I suspect they are wrong and we should fix them.

Reply:
I'm also getting a bit confused now on the `env`. Now you saw that the `this.system` is a `env`. And `env` has a `name` property as in the `lookupEnvs`. You can see the value of `this.system` here.

https://github.com/oppia/oppia/blob/13b1da20ee6c0e4eabc9720a3d1ca3d87c62fe8c/core/templates/expressions/expression-syntax-tree.service.ts#L158-L175

Now this is a totally different thing?

Reply:
Isn't `name` an argument, not a `property`? It can take different values, it's not the literal string "name".

Reply:
Oh, yes sorry. But I have one more question. 
```typescript
<string> env[name];
```
suggests that the properties in env should be string. But looking at `this.system` above we don't see that ?

Reply:
Yes, that seems wrong.

Reply:
Wait actually I'm not 100% certain about that. Let me read it a bit more. It would help if you could also try to understand what's going on too.

Reply:
OK, interesting, this might be polymorphic. I can't do this right now (in meetings) but what I suggest you do is take a look at core/templates/expressions/expression-type-parser.service.spec.ts and try to puzzle out how that works, especially around the calls to applyFunctionToParseTree(). (But take a look at the whole test to understand what it's doing.) See if you can post your analysis/understanding here regarding the role of this.system, and if you need help after that, we can figure out next steps.

Reply:
As far as I understood the type of env from it's usage like

https://github.com/oppia/oppia/blob/38e07878998a252835fcdce895fe8e06f6124d47/core/templates/expressions/expression-type-parser.service.ts#L62-L63

That the type of env will be like `this.system` i.e. two functions `eval and getType`.

Reply:
Can you provide more analysis? What are other examples of envs besides `this.system`? Point to these in the codebase at the points where it's specified, try to "show your work" so that the reader can understand how you got to your conclusion.

Try to do this for all questions, otherwise we keep going back and forth.

Reply:
Ok. So first of all `ExpressionSyntaxTreeService` is used in three places.

- `ExpressionTypeParserService` - This service actually isn't used anywhere. But uses only `lookupEnvs` function.
- `ExpressionEvaluatorService` - This service is used in `ExpressionInterpolationService` and uses `lookupEnvs` and `applyFunctionToParseTree` function.
- `ExpressionInterpolationService` - Uses both `ExpressionSyntaxTreeService` and `ExpressionEvaluatorService`. Uses only `getParamsUsedInExpression` function of `ExpressionSyntaxTreeService` which doesn't have anything to do with `envs`. This is the **only** service that is used in other places in the codebase.

Now only the `evaluateExpression` function in the `ExpressionEvaluatorService` is used in the `ExpressionInterpolationService`. So, the main function that could tell what's really happening is `evaluateExpression`.

https://github.com/oppia/oppia/blob/38e07878998a252835fcdce895fe8e06f6124d47/core/templates/expressions/expression-evaluator.service.ts#L72-L75

As it can be seen that it calls `applyFunctionToParseTree` which is 
https://github.com/oppia/oppia/blob/38e07878998a252835fcdce895fe8e06f6124d47/core/templates/expressions/expression-syntax-tree.service.ts#L77-L80

Which concats the `this.system` to the envs. Now finally the `func` argument in the above function always is `evaluate` function which is.

https://github.com/oppia/oppia/blob/38e07878998a252835fcdce895fe8e06f6124d47/core/templates/expressions/expression-evaluator.service.ts#L84-L109

This finally clears all the doubts regarding the envs.

So initially envs are just the values of external variables like `a` or `b` and looks something like
```typescript
[{
  a: 1,
  b: 2
}]
```
and afterwards the `this.system` is also included in the envs.

and when there is a `expression` passed to `evaluateExpression` like `a+b`. The `ExpressionParserService.parse` converts it to something like `['+', ['#', 'a'], ['#', 'b']]`. `#` as I've understood is something which tells that the next element is a variable. And hence the `evaluate` function returns the result `a+b` i.e. 3. 

## Conclusion

Env can be either `this.system` or the values of variables. (The exploration parameters which I'm assuming would be string or numbers). I've updated the types accordingly.

Reply:
Good, well done, this sounds right to me. Thanks @nishantwrp! (And hopefully this provides a blueprint for the sort of thing you should do when you conduct future investigations :) )